File: go/nbh_vec6.go
/******************************************************************************
 *                                  LICENSE                                   *
 ******************************************************************************
 *  This file is part of newtonian_black_holes.                               *
 *                                                                            *
 *  newtonian_black_holes is free software: you can redistribute it and/or    *
 *  modify it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation, either version 3 of the License, or      *
 *  (at your option) any later version.                                       *
 *                                                                            *
 *  newtonian_black_holes is distributed in the hope that it will be useful   *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 *  GNU General Public License for more details.                              *
 *                                                                            *
 *  You should have received a copy of the GNU General Public License         *
 *  along with newtonian_black_holes.  If not, see                            *
 *  <https://www.gnu.org/licenses/>.                                          *
 ******************************************************************************
 *  Purpose:                                                                  *
 *      Provides a basic 6D vector struct to represent "phase-space."         *
 ******************************************************************************
 *  Author: Ryan Maguire                                                      *
 *  Date:   2023/09/23                                                        *
 ******************************************************************************/
package main

/*  Sqrt function provided here, used for calculating norms.                  */
import "math"

/*  Struct for working in phase space. Points in R^6 are two points in R^3.   */
type Vec6 struct {

    /*  A point in phase-space is defined by its position and velocity.       */
    P, V Vec3
}

/******************************************************************************
 *                         Vec6 Functions and Methods                         *
 ******************************************************************************/

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6FromReals                                                         *
 *  Purpose:                                                                  *
 *      Creates a 6D vector from "rectangular", or Cartesian, coordinates.    *
 *  Arguments:                                                                *
 *      x (float64):                                                          *
 *          The x-component of the position vector.                           *
 *      y (float64):                                                          *
 *          The y-component of the position vector.                           *
 *      z (float64):                                                          *
 *          The z-component of the position vector.                           *
 *      vx (float64):                                                         *
 *          The x-component of the velocity vector.                           *
 *      vy (float64):                                                         *
 *          The y-component of the velocity vector.                           *
 *      vz (float64):                                                         *
 *          The z-component of the velocity vector.                           *
 *  Outputs:                                                                  *
 *      u (Vec6):                                                             *
 *          The vector (x, y, z, vx, vy, vz).                                 *
 *  Method:                                                                   *
 *      Set the individual components for the struct and return.              *
 ******************************************************************************/
func Vec6FromReals(x, y, z, vx, vy, vz float64) {
    var u Vec6
    u.P.X = x
    u.P.Y = y
    u.P.Z = z
    u.V.X = vx
    u.V.Y = vy
    u.V.Z = vz
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6Add                                                               *
 *  Purpose:                                                                  *
 *      Adds two vectors in R^6.                                              *
 *  Arguments:                                                                *
 *      u0 (*Vec6):                                                           *
 *          A pointer to a vector.                                            *
 *      u1 (*Vec6):                                                           *
 *          Another pointer to a vector.                                      *
 *  Outputs:                                                                  *
 *      sum (Vec6):                                                           *
 *          The vector sum of u0 and u1.                                      *
 *  Method:                                                                   *
 *      Vector addition is performed component-wise. That is, given two       *
 *      vectors u0 = (p0, v0) and u1 = (p1, v1), the sum is:                  *
 *                                                                            *
 *          sum = u0 + u1                                                     *
 *              = (p0, v0) + (p1, v1)                                         *
 *              = (p0 + p1, v0 + v1)                                          *
 *                                                                            *
 *      This is computed and the sum is returned.                             *
 ******************************************************************************/
func Vec6Add(u0, u1 *Vec6) Vec6 {
    return Vec6{Vec3Add(&u0.P, &u1.P), Vec3Add(&u0.V, &u1.V)}
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6AddTo                                                             *
 *  Purpose:                                                                  *
 *      Adds two vectors in R^6 and stores the result in the first variable.  *
 *  Arguments:                                                                *
 *      u0 (*Vec6):                                                           *
 *          A pointer to a vector. The sum is stored here.                    *
 *      u1 (*Vec6):                                                           *
 *          Another pointer to a vector.                                      *
 *  Outputs:                                                                  *
 *      None.                                                                 *
 *  Method:                                                                   *
 *      Perform vector addition component-wise and store the result in u0.    *
 ******************************************************************************/
func Vec6AddTo(u0, u1 *Vec6) {
    Vec3AddTo(&u0.P, &u1.P)
    Vec3AddTo(&u0.V, &u1.V)
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6Subtract                                                          *
 *  Purpose:                                                                  *
 *      Subtracts two vectors in R^6.                                         *
 *  Arguments:                                                                *
 *      u0 (*Vec6):                                                           *
 *          A pointer to a vector.                                            *
 *      u1 (*Vec6):                                                           *
 *          Another pointer to a vector.                                      *
 *  Outputs:                                                                  *
 *      diff (Vec6):                                                          *
 *          The vector difference of u0 and u1, u0 - u1.                      *
 *  Method:                                                                   *
 *      Vector subtraction is performed component-wise. That is, given two    *
 *      vectors u0 = (p0, v0) and u1 = (p1, v1), the difference is:           *
 *                                                                            *
 *          diff = u0 - u1                                                    *
 *               = (p0, v0) - (p1, v1)                                        *
 *               = (p0 - p1, v0 - v1)                                         *
 *                                                                            *
 *      This is computed and the difference is returned.                      *
 *  Notes:                                                                    *
 *      Vector subtraction is not commutative. Vec6Subtract(&u0, &u1)         *
 *      computes the difference u0 - u1 (and not u1 - u0).                    *
 ******************************************************************************/
func Vec6Subtract(u0, u1 *Vec6) Vec6 {
    return Vec6{Vec3Subtract(&u0.P, &u1.P), Vec3Subtract(&u0.V, &u1.V)}
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6SubtractFrom                                                      *
 *  Purpose:                                                                  *
 *      Subtracts two vectors and stores the result in the first variable.    *
 *  Arguments:                                                                *
 *      u0 (*Vec6):                                                           *
 *          A pointer to a vector. The difference is stored here.             *
 *      u1 (*Vec6):                                                           *
 *          Another pointer to a vector.                                      *
 *  Outputs:                                                                  *
 *      None.                                                                 *
 *  Method:                                                                   *
 *      Perform vector subtraction component-wise and store the result in u0. *
 *  Notes:                                                                    *
 *      Vector subtraction is not commutative. Vec6SubtractFrom(&u0, &u1)     *
 *      computes the difference u0 - u1 (and not u1 - u0).                    *
 ******************************************************************************/
func Vec6SubtractFrom(u0, u1 *Vec6) {
    Vec3SubtractFrom(&u0.P, &u1.P)
    Vec3SubtractFrom(&u0.V, &u1.V)
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6Scale                                                             *
 *  Purpose:                                                                  *
 *      Performs scalar multiplication.                                       *
 *  Arguments:                                                                *
 *      a (float64):                                                          *
 *          A real number.                                                    *
 *      u (*Vec6):                                                            *
 *          A pointer to a vector.                                            *
 *  Outputs:                                                                  *
 *      prod (Vec6):                                                          *
 *          The scalar product a*u.                                           *
 *  Method:                                                                   *
 *      Scalar multiplication is performed component-wise. That is, if u is   *
 *      the vector u = (p, v), then the product is:                           *
 *                                                                            *
 *          prod = a*u                                                        *
 *               = a*(p, v)                                                   *
 *               = (a*p, a*v)                                                 *
 *                                                                            *
 *      This is computed and the product is returned.                         *
 ******************************************************************************/
func Vec6Scale(a float64, u *Vec6) Vec6 {
    return Vec6{Vec3Scale(a, &u.P), Vec3Scale(a, &u.V)}
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6ScaleBy                                                           *
 *  Purpose:                                                                  *
 *      Performs scalar multiplication and stores the result in place.        *
 *  Arguments:                                                                *
 *      a (float64):                                                          *
 *          A real number.                                                    *
 *      u (*Vec6):                                                            *
 *          A pointer to a vector. The product is stored here.                *
 *  Outputs:                                                                  *
 *      None.                                                                 *
 *  Method:                                                                   *
 *      Perform scalar multiplication component-wise, storing the result in u.*
 ******************************************************************************/
func Vec6ScaleBy(a float64, u *Vec6) {
    Vec3ScaleBy(a, &u.P)
    Vec3ScaleBy(a, &u.V)
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6.Dot                                                              *
 *  Purpose:                                                                  *
 *      Performs the Euclidean dot product in R^6.                            *
 *  Arguments:                                                                *
 *      u1 (*Vec6):                                                           *
 *          Another pointer to a vector.                                      *
 *  Outputs:                                                                  *
 *      dot (float64):                                                        *
 *          The dot product u0 . u1.                                          *
 *  Method:                                                                   *
 *      The Euclidean dot product sums the products of the components. Given  *
 *      two vectors u0 = (p0, v0) and u1 = (p1, v1), the dot product is:      *
 *                                                                            *
 *          u0 . u1 = (p0, v0) . (p1, v1)                                     *
 *                  = (p0 . p1) + (v0 . v1)                                   *
 *                                                                            *
 *      This is computed and the sum is returned.                             *
 ******************************************************************************/
func (u0 *Vec6) Dot(u1 *Vec6) float64 {
    return u0.P.Dot(&u1.P) + u0.V.Dot(&u1.V)
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6.NormSq                                                           *
 *  Purpose:                                                                  *
 *      Computes the square of the Euclidean, or L2, norm for vectors in R^6. *
 *  Arguments:                                                                *
 *      None.                                                                 *
 *  Outputs:                                                                  *
 *      normsq_u (float64):                                                   *
 *          The square of the norm, or magnitude, or length of the vector u.  *
 *  Method:                                                                   *
 *      Use the Pythagorean theorem to compute. Given u = (p, v) we have:     *
 *                                                                            *
 *          ||u||^2 = ||(p, v)||^2                                            *
 *                  = ||p||^2 + ||v||^2                                       *
 *                                                                            *
 *      Compute and return.                                                   *
 ******************************************************************************/
func (u *Vec6) NormSq() float64 {
    return u.P.NormSq() + u.V.NormSq()
}

/******************************************************************************
 *  Function:                                                                 *
 *      Vec6.Norm                                                             *
 *  Purpose:                                                                  *
 *      Computes the Euclidean, or L2, norm for vectors in R^6.               *
 *  Arguments:                                                                *
 *      None.                                                                 *
 *  Outputs:                                                                  *
 *      norm_u (float64):                                                     *
 *          The norm, or magnitude, or length of the vector u.                *
 *  Method:                                                                   *
 *      Return the square root of the Vec6.NormSq function.                   *
 ******************************************************************************/
func (u *Vec6) Norm() float64 {
    return math.Sqrt(u.NormSq())
}

File: pascal/nbh_vec6.pas
(******************************************************************************
 *                                  LICENSE                                   *
 ******************************************************************************
 *  This file is part of newtonian_black_holes.                               *
 *                                                                            *
 *  newtonian_black_holes is free software: you can redistribute it and/or    *
 *  modify it under the terms of the GNU General Public License as published  *
 *  by the Free Software Foundation, either version 3 of the License, or      *
 *  (at your option) any later version.                                       *
 *                                                                            *
 *  newtonian_black_holes is distributed in the hope that it will be useful   *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 *  GNU General Public License for more details.                              *
 *                                                                            *
 *  You should have received a copy of the GNU General Public License         *
 *  along with newtonian_black_holes.  If not, see                            *
 *  <https://www.gnu.org/licenses/>.                                          *
 ******************************************************************************)
UNIT nbh_Vec6;

INTERFACE

(*  3D vectors are used to represent 6D vectors using "phase-space."          *)
USES nbh_Vec3;

TYPE
    (*  A 6D vector consists of two 3D vectors: Position and velocity.        *)
    Vec6EnumType = 0..1;
    Vec6 = Array[Vec6EnumType] of Vec3;

(*  Computes a 6D vector from two 3D vectors, the position and velocity.      *)
Function Vec6FromVectors(const P, V: Vec3) : Vec6;

(*  Computes a 6D vector from 6 real numbers, the individual components.      *)
Function Vec6FromReals(const x, y, z, vx, vy, vz: Real) : Vec6;

(*  Performs vector addition for two vectors in R^6.                          *)
Function Vec6Add(const U0, U1: Vec6) : Vec6;

(*  Performs vector addition in R^6 and stores the result in U0.              *)
Procedure Vec6AddTo(var U0: Vec6; const U1: Vec6);

(*  Performs vector subtraction for two vectors in R^6.                       *)
Function Vec6Subtract(const U0, U1: Vec6) : Vec6;

(*  Performs vector subtraction and stores the result in U0.                  *)
Procedure Vec6SubtractFrom(var U0: Vec6; const U1: Vec6);

(*  Performs scalar multiplication for 6D vectors.                            *)
Function Vec6Scale(const a: Real; const U: Vec6) : Vec6;

(*  Performs scalar multiplication and stores the result in U.                *)
Procedure Vec6ScaleBy(const a: Real; var U: Vec6);

(*  Performs the Euclidean dot product in R^6.                                *)
Function Vec6DotProduct(const U0, U1: Vec6) : Real;

(*  Computes the square of the Euclidean norm for vectors in R^6.             *)
Function Vec6NormSq(const U: Vec6) : Real;

(*  Computes the magnitude, or length, of vectors in R^6.                     *)
Function Vec6Norm(const U: Vec6) : Real;

IMPLEMENTATION

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6FromVectors                                                       *
 *  Purpose:                                                                  *
 *      Returns a 6D vector from the specified position and velocity vectors. *
 *  Arguments:                                                                *
 *      P (const Vec3):                                                       *
 *          The position vector.                                              *
 *      V (const Vec3):                                                       *
 *          The velocity vector.                                              *
 *  Output:                                                                   *
 *      Vec6FromVectors (Vec6):                                               *
 *          The vector (P, V).                                                *
 ******************************************************************************)
Function Vec6FromVectors(const P, V: Vec3) : Vec6;
BEGIN
    Vec6FromVectors[0] := P;
    Vec6FromVectors[1] := V;
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6FromReals                                                         *
 *  Purpose:                                                                  *
 *      Returns a 6D vector from the specified Cartesian coordinates.         *
 *  Arguments:                                                                *
 *      x (const Real):                                                       *
 *          The x-coordinate of the position vector.                          *
 *      y (const Real):                                                       *
 *          The y-coordinate of the position vector.                          *
 *      z (const Real):                                                       *
 *          The z-coordinate of the position vector.                          *
 *      vx (const Real):                                                      *
 *          The x-coordinate of the velocity vector.                          *
 *      vy (const Real):                                                      *
 *          The y-coordinate of the velocity vector.                          *
 *      vz (const Real):                                                      *
 *          The z-coordinate of the velocity vector.                          *
 *  Output:                                                                   *
 *      Vec6FromReals (Vec6):                                                 *
 *          The vector (x, y, z, vx, vy, vz).                                 *
 ******************************************************************************)
Function Vec6FromReals(const x, y, z, vx, vy, vz: Real) : Vec6;
VAR
    (*  Two variables representing the position and velocity components.      *)
    P, V: Vec3;

BEGIN
    (*  Compute the individual position and velocity components and return.   *)
    P := Vec3Rect(x, y, z);
    V := Vec3Rect(vx, vy, vz);
    Vec6FromReals := Vec6FromVectors(P, V);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6Add                                                               *
 *  Purpose:                                                                  *
 *      Performs vector addition in R^6.                                      *
 *  Arguments:                                                                *
 *      U0 (const Vec6):                                                      *
 *          A vector in R^6.                                                  *
 *      U1 (const Vec6):                                                      *
 *          Another vector in R^6.                                            *
 *  Output:                                                                   *
 *      Vec6Add (Vec6):                                                       *
 *          The vector sum U0 + U1.                                           *
 ******************************************************************************)
Function Vec6Add(const U0, U1: Vec6) : Vec6;
BEGIN
    Vec6Add[0] := Vec3Add(U0[0], U1[0]);
    Vec6Add[1] := Vec3Add(U0[1], U1[1]);
END;

(******************************************************************************
 *  Procedure:                                                                *
 *      Vec6AddTo                                                             *
 *  Purpose:                                                                  *
 *      Performs vector addition in R^6.                                      *
 *  Arguments:                                                                *
 *      U0 (var Vec6):                                                        *
 *          A vector in R^6. The sum is stored here.                          *
 *      U1 (const Vec6):                                                      *
 *          Another vector in R^6.                                            *
 ******************************************************************************)
Procedure Vec6AddTo(var U0: Vec6; const U1: Vec6);
BEGIN
    Vec3AddTo(U0[0], U1[0]);
    Vec3AddTo(U0[1], U1[1]);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec36ubtract                                                          *
 *  Purpose:                                                                  *
 *      Performs vector subtraction in R^6.                                   *
 *  Arguments:                                                                *
 *      U0 (const Vec6):                                                      *
 *          A vector in R^6.                                                  *
 *      U1 (const Vec6):                                                      *
 *          Another vector in R^6.                                            *
 *  Output:                                                                   *
 *      Vec6Subtract (Vec6):                                                  *
 *          The vector difference U0 - U1.                                    *
 ******************************************************************************)
Function Vec6Subtract(const U0, U1: Vec6) : Vec6;
BEGIN
    Vec6Subtract[0] := Vec3Subtract(U0[0], U1[0]);
    Vec6Subtract[1] := Vec3Subtract(U0[1], U1[1]);
END;

(******************************************************************************
 *  Procedure:                                                                *
 *      Vec6SubtractFrom                                                      *
 *  Purpose:                                                                  *
 *      Performs vector subtraction in R^6.                                   *
 *  Arguments:                                                                *
 *      U0 (var Vec6):                                                        *
 *          A vector in R^6. The difference is stored here.                   *
 *      U1 (const Vec6):                                                      *
 *          Another vector in R^6.                                            *
 ******************************************************************************)
Procedure Vec6SubtractFrom(var U0: Vec6; const U1: Vec6);
BEGIN
    Vec3SubtractFrom(U0[0], U1[0]);
    Vec3SubtractFrom(U0[1], U1[1]);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6Scale                                                             *
 *  Purpose:                                                                  *
 *      Performs scalar multiplication for vectors in R^6.                    *
 *  Arguments:                                                                *
 *      a (const Real):                                                       *
 *          A real number, the scalar multiplier.                             *
 *      U (const Vec6):                                                       *
 *          A vector in R^6.                                                  *
 *  Output:                                                                   *
 *      Vec6Scale (Vec6):                                                     *
 *          The scalar product a*U.                                           *
 ******************************************************************************)
Function Vec6Scale(const a: Real; const U: Vec6) : Vec6;
BEGIN
    Vec6Scale[0] := Vec3Scale(a, U[0]);
    Vec6Scale[1] := Vec3Scale(a, U[1]);
END;

(******************************************************************************
 *  Procedure:                                                                *
 *      Vec6ScaleBy                                                           *
 *  Purpose:                                                                  *
 *      Performs scalar multiplication and stores the result in U.            *
 *  Arguments:                                                                *
 *      a (const Real):                                                       *
 *          A real number, the scalar multiplier.                             *
 *      U (var Vec6):                                                         *
 *          A vector in R^6. The product is stored here.                      *
 ******************************************************************************)
Procedure Vec6ScaleBy(const a: Real; var U: Vec6);
BEGIN
    Vec3ScaleBy(a, U[0]);
    Vec3ScaleBy(a, U[1]);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6DotProduct                                                        *
 *  Purpose:                                                                  *
 *      Computes the Euclidean dot product of two vectors in R^6.             *
 *  Arguments:                                                                *
 *      U0 (const Vec6):                                                      *
 *          A vector in R^6.                                                  *
 *      U1 (const Vec6):                                                      *
 *          Another vector in R^6.                                            *
 *  Output:                                                                   *
 *      Vec6DotProduct (Real):                                                *
 *          The Euclidean dot product U0 . U1.                                *
 ******************************************************************************)
Function Vec6DotProduct(const U0, U1: Vec6) : Real;
BEGIN
    Vec6DotProduct := Vec3DotProduct(U0[0], U1[0]) +
                      Vec3DotProduct(U0[1], U1[1]);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6NormSq                                                            *
 *  Purpose:                                                                  *
 *      Computes the square of the Euclidean norm of a vector in R^6.         *
 *  Arguments:                                                                *
 *      U (const Vec6):                                                       *
 *          A vector in R^6.                                                  *
 *  Output:                                                                   *
 *      Vec6NormSq (Real):                                                    *
 *          The square of the Euclidean norm of U, ||U||^2.                   *
 ******************************************************************************)
Function Vec6NormSq(const U: Vec6) : Real;
BEGIN
    Vec6NormSq := Vec3NormSq(U[0]) + Vec3NormSq(U[1]);
END;

(******************************************************************************
 *  Function:                                                                 *
 *      Vec6Norm                                                              *
 *  Purpose:                                                                  *
 *      Computes the Euclidean norm of a vector in R^6.                       *
 *  Arguments:                                                                *
 *      U (const Vec6):                                                       *
 *          A vector in R^6.                                                  *
 *  Output:                                                                   *
 *      Vec3Norm (Real):                                                      *
 *          The Euclidean norm of U, ||U||.                                   *
 ******************************************************************************)
Function Vec6Norm(const U: Vec6) : Real;
BEGIN
    Vec6Norm := Sqrt(Vec6NormSq(U));
END;

END.
